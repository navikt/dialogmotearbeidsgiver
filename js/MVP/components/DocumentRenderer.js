import React from 'react';
import PropTypes from 'prop-types';
import { Element, Innholdstittel, Normaltekst } from 'nav-frontend-typografi';
import Lenke from 'nav-frontend-lenker';

const DocumentTypes = {
  HEADER: 'HEADER',
  PARAGRAPH: 'PARAGRAPH',
  LINK: 'LINK',
};

const DocumentRenderer = ({ documentComponent }) => {
  const { type, title, texts } = documentComponent;

  switch (type) {
    case DocumentTypes.HEADER:
      return (
        <React.Fragment>
          {texts.map((text, index) => (
            <Innholdstittel key={index}>{text}</Innholdstittel>
          ))}
        </React.Fragment>
      );

    case DocumentTypes.LINK:
      return (
        <React.Fragment>
          {title && <Element>{title}</Element>}
          {texts.map((text, index) => (
            <Lenke key={index} href={text}>
              {text}
            </Lenke>
          ))}
        </React.Fragment>
      );

    case DocumentTypes.PARAGRAPH:
      return (
        <React.Fragment>
          {title && <Element>{title}</Element>}
          {texts.map((text, index) => (
            <Normaltekst key={index}>{text}</Normaltekst>
          ))}
        </React.Fragment>
      );

    default:
      return null;
  }
};

DocumentRenderer.propTypes = {
  documentComponent: PropTypes.shape({
    type: PropTypes.oneOf(Object.values(DocumentTypes)),
    title: PropTypes.string,
    texts: PropTypes.arrayOf(PropTypes.string),
  }),
};

export default DocumentRenderer;
