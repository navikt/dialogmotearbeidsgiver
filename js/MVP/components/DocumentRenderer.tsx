import React from 'react';
import { Element, Innholdstittel, Normaltekst } from 'nav-frontend-typografi';
import { DocumentTypes } from '../globals/constants';
import Lenke from 'nav-frontend-lenker';
import { trackOnClick } from '@/amplitude/amplitude';
import { eventNames } from '@/amplitude/events';
import { BrevDocument } from '@/api/types/brevTypes';

interface DocumentRendererProps {
  documentComponent: BrevDocument;
}

const DocumentRenderer = ({ documentComponent }: DocumentRendererProps) => {
  const { type, title, texts } = documentComponent;

  switch (type) {
    case DocumentTypes.HEADER:
      return (
        <React.Fragment>
          {texts.map((text, index) => (
            <Innholdstittel key={index}>{text}</Innholdstittel>
          ))}
        </React.Fragment>
      );

    case DocumentTypes.LINK:
      return (
        <React.Fragment>
          {title && <Element>{title}</Element>}
          {texts.map((text, index) => (
            <Lenke
              key={index}
              href={text}
              onClick={() => trackOnClick(eventNames.documentRendererLink, { linkType: text })}
            >
              {text}
            </Lenke>
          ))}
        </React.Fragment>
      );

    case DocumentTypes.PARAGRAPH:
      return (
        <React.Fragment>
          {title && <Element>{title}</Element>}
          {texts.map((text, index) => (
            <Normaltekst key={index}>{text}</Normaltekst>
          ))}
        </React.Fragment>
      );

    default:
      return null;
  }
};

export default DocumentRenderer;
