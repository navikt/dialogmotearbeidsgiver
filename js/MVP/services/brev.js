import { get, post, getRaw } from '../../gateway-api/gatewayApi';
import { ISDIALOGMOTE_PROXY_BASE_PATH } from '../globals/paths';

export const postLestBrev = async (uuid) => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/les`;
  return post(url);
};

export const getBrev = async (fnr) => {
  const url = ISDIALOGMOTE_PROXY_BASE_PATH;
  const header = { 'nav-personident': fnr };
  return get(url, header);
};

export const getBrevPdf = async (uuid) => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/pdf`;
  return getRaw(url);
};
