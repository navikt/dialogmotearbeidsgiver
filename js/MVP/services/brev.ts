import { ISDIALOGMOTE_PROXY_BASE_PATH } from '../globals/paths';
import { get, post } from '@/api/axios';
import { Brev } from '@/api/types/brevTypes';

export const postLestBrev = (uuid: string): Promise<void> => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/les`;
  return post(url);
};

export const getBrev = (fnr?: string): Promise<Brev[]> => {
  return get(ISDIALOGMOTE_PROXY_BASE_PATH, { personIdent: fnr });
};

export const getBrevPdf = (uuid: string): Promise<Blob> => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/pdf`;
  return get(url, { responseType: 'blob' });
};
