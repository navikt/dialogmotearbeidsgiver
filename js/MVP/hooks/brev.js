import { useMutation, useQuery, useQueryClient } from 'react-query';
import { getBrev, postLestBrev } from '../services/brev';

const MOTEINNKALLELSE = 'moteinnkallelse';

export const useBrev = () => {
  return useQuery(MOTEINNKALLELSE, getBrev, {
    retry: 0,
  });
};

const setLestDatoForBrev = (uuid) => {
  return (varsel) => {
    if (varsel.uuid === uuid) {
      return { ...varsel, lestDato: true };
    }

    return varsel;
  };
};

export const useMutateBrevLest = () => {
  const queryClient = useQueryClient();

  return useMutation(({ uuid }) => postLestBrev(uuid), {
    onSuccess: (_data, variables) => {
      queryClient.setQueryData([MOTEINNKALLELSE], (old) => {
        return old.map(setLestDatoForBrev(variables.uuid));
      });
    },
  });
};
