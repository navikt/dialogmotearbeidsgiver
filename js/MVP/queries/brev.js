import { useMutation, useQuery, useQueryClient } from 'react-query';
import { getBrev, postLestBrev } from '../services/brev';
import { useSykmeldt } from './sykmeldt';

const brevQueryKeys = (koblingId) => ['brev', koblingId];

export const useBrev = (koblingId) => {
  const { fnr } = useSykmeldt(koblingId);
  return useQuery(brevQueryKeys(koblingId), () => getBrev(fnr), {
    enabled: !!fnr,
  });
};

const setLestDatoForBrev = (uuid) => {
  return (varsel) => {
    if (varsel.uuid === uuid) {
      return { ...varsel, lestDato: true };
    }

    return varsel;
  };
};

export const useMutateBrevLest = () => {
  const queryClient = useQueryClient();

  return useMutation(({ brevUuid }) => postLestBrev(brevUuid), {
    onMutate: (variables) => {
      queryClient.setQueryData(brevQueryKeys(variables.koblingId), (old) => {
        return old.map(setLestDatoForBrev(variables.brevUuid));
      });
    },
    onSettled: (variables) => {
      queryClient.invalidateQueries(brevQueryKeys(variables.koblingId));
    },
  });
};
