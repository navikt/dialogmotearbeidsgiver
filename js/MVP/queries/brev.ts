import { useMutation, useQuery, useQueryClient } from 'react-query';
import { getBrev, postLestBrev } from '../services/brev';
import { useSykmeldte } from './sykmeldte';

const brevQueryKeys = (narmestelederId) => ['brev', narmestelederId];

export const useBrev = (narmestelederId: string) => {
  const sykmeldteQuery = useSykmeldte(narmestelederId);
  const fnr = sykmeldteQuery.isSuccess ? sykmeldteQuery.data.fnr : null;
  return useQuery(brevQueryKeys(narmestelederId), () => getBrev(fnr), {
    enabled: sykmeldteQuery.isSuccess && !!sykmeldteQuery.data.fnr,
  });
};

const setLestDatoForBrev = (uuid: string) => {
  return (varsel) => {
    if (varsel.uuid === uuid) {
      return { ...varsel, lestDato: new Date().toString() };
    }

    return varsel;
  };
};

export const useMutateBrevLest = () => {
  const queryClient = useQueryClient();

  return useMutation(({ brevUuid }) => postLestBrev(brevUuid), {
    onMutate: (variables: { narmestelederId: string; brevUuid: string }) => {
      queryClient.setQueryData(brevQueryKeys(variables.narmestelederId), (old: string[]) => {
        return old.map(setLestDatoForBrev(variables.brevUuid));
      });
    },
    onSettled: (variables: { narmestelederId: string; brevUuid: string }) => {
      queryClient.invalidateQueries(brevQueryKeys(variables.narmestelederId));
    },
  });
};
