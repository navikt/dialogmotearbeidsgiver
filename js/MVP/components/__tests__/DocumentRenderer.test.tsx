import { render } from '@testing-library/react';
import DocumentRenderer from '../DocumentRenderer';
import React from 'react';
import { createBrevDocument } from '@/MVP/test/fixtures/brev';

describe('DocumentRenderer', () => {
  test('should render PARAGRAPH', async () => {
    const doc = createBrevDocument();

    const { container } = render(<DocumentRenderer documentComponent={doc} />);

    expect(container).toMatchSnapshot();
  });

  test('should render HEADER', async () => {
    const doc = createBrevDocument({ type: 'HEADER' });

    const { container } = render(<DocumentRenderer documentComponent={doc} />);

    expect(container).toMatchSnapshot();
  });

  test('should render LINK', async () => {
    const doc = createBrevDocument({ type: 'LINK' });

    const { container } = render(<DocumentRenderer documentComponent={doc} />);

    expect(container).toMatchSnapshot();
  });

  test('should not render', async () => {
    const doc = createBrevDocument({ type: undefined });

    const { container } = render(<DocumentRenderer documentComponent={doc} />);

    expect(container).toMatchSnapshot();
  });
});
