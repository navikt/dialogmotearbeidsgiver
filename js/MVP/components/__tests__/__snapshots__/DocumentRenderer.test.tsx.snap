// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentRenderer should not render 1`] = `<div />`;

exports[`DocumentRenderer should render HEADER 1`] = `
<div>
  <h1
    class="typo-innholdstittel"
  >
    Test_text_1
  </h1>
</div>
`;

exports[`DocumentRenderer should render LINK 1`] = `
<div>
  <p
    class="typo-element"
  >
    TEST_HEADER
  </p>
  <a
    class="lenke"
    href="Test_text_1"
  >
    Test_text_1
  </a>
</div>
`;

exports[`DocumentRenderer should render PARAGRAPH 1`] = `
<div>
  <p
    class="typo-element"
  >
    TEST_HEADER
  </p>
  <p
    class="typo-normal"
  >
    Test_text_1
  </p>
</div>
`;
